fendermint/vm/interpreter/src/genesis.rs

```diff
@@ -0,0 +1,795 @@
+    pub fn deploy_contract<T>(
+        &self,
+        state: &mut FvmGenesisState<DB>,
+        contract_name: &str,
+        constructor_params: T,
+    ) -> anyhow::Result<et::Address>
+    where
+        T: Tokenize,
+    {
+        let contract = self.top_contract(contract_name)?;
+        let contract_id = contract.actor_id;
+        let contract_src = contract_src(contract_name);
+
+        let bytecode = self
+            .hardhat
+            .bytecode(contract_src, contract_name, &self.lib_addrs)
+            .with_context(|| format!("failed to load {contract_name} bytecode"))?;
+
+        let eth_addr = state
+            .create_evm_actor_with_cons(contract_id, &contract.abi, bytecode, constructor_params)
+            .with_context(|| format!("failed to create {contract_name} actor"))?;
+
+        let id_addr = et::Address::from(EthAddress::from_id(contract_id).0);
+        let eth_addr = et::Address::from(eth_addr.0);
+
+        tracing::info!(
+            actor_id = contract_id,
+            ?eth_addr,
+            ?id_addr,
+            contract_name,
+            "deployed Ethereum contract"
+        );
+
+        // The Ethereum address is more usable inside the EVM than the ID address.
+        Ok(eth_addr)
+    }
+
+    /// Collect Facet Cuts for the diamond pattern, where the facet address comes from already deployed library facets.
+    pub fn facets(&self, contract_name: &str) -> anyhow::Result<Vec<FacetCut>> {
+        let contract = self.top_contract(contract_name)?;
+        let mut facet_cuts = Vec::new();
+
+        for facet in contract.facets.iter() {
+            let facet_name = facet.name;
+            let facet_src = contract_src(facet_name);
+            let facet_fqn = self.hardhat.fqn(&facet_src, facet_name);
+
+            let facet_addr = self
+                .lib_addrs
+                .get(&facet_fqn)
+                .ok_or_else(|| anyhow!("facet {facet_name} has not been deployed"))?;
+
+            let method_sigs = facet
+                .abi
+                .functions()
+                .filter(|f| f.signature() != "init(bytes)")
+                .map(|f| f.short_signature())
+                .collect();
+
+            let facet_cut = FacetCut {
+                facet_address: *facet_addr,
+                action: 0, // Add
+                function_selectors: method_sigs,
+            };
+
+            facet_cuts.push(facet_cut);
+        }
+
+        Ok(facet_cuts)
+    }
+
+    fn top_contract(&self, contract_name: &str) -> anyhow::Result<&EthContract> {
+        self.top_contracts
+            .get(contract_name)
+            .ok_or_else(|| anyhow!("unknown top contract name: {contract_name}"))
+    }
+}
+
+/// Sum of balances in the genesis accounts.
+fn circ_supply(g: &Genesis) -> TokenAmount {
+    g.accounts
+        .iter()
+        .fold(TokenAmount::zero(), |s, a| s + a.balance.clone())
+}
+//
+// #[cfg(test)]
```

Comment:
We probably want to either re-write the test so that they work with the new implementation or just delete them at least?

Reply:
I will remove them first, maybe rewrite them later